torch/csrc/jit/script/compiler.cpp

```diff
@@ -1298,246 +1302,92 @@ struct to_ir {
-          {},
-          range);
-    }
-
-    auto sizes_tuple = emitBuiltinCall(
-        range,
-        *graph,
-        aten::size,
-        c10::nullopt,
-        {tensorArg},
-        {},
-        /*required=*/true);
-
-    auto max_trip_count_val = emitBuiltinCall(
-        range,
-        *graph,
-        aten::select,
-        c10::nullopt,
-        {sizes_tuple, outermost_dim_index},
-        {},
-        /*required=*/true);
-
-    const auto& ident_name = target.name();
-    auto assigner = [outermost_dim_index, ident_name, range, tensorArg, this](
-                        Value* index, std::shared_ptr<Environment> env) {
-      auto cur_elm = emitBuiltinCall(
-          range,
-          *this->graph,
-          aten::select,
-          c10::nullopt,
-          {tensorArg, outermost_dim_index, index},
-          {},
-          /*required=*/true);
-      env->setVar(range, ident_name, cur_elm);
-    };
-
-    emitLoopCommon(range, body, assigner, {}, max_trip_count_val);
   }
 
   void emitFor(const For& stmt) {
-    // For now, we only support range loops. e.g. for i in range(3): ...
-
     auto targets = stmt.targets();
     auto itrs = stmt.itrs();
     auto body = stmt.body();
     if (stmt.itrs().size() != 1) {
       throw ErrorReport(stmt)
           << "List of iterables is not supported currently.";
     }
+
     if (targets.size() != 1) {
       throw ErrorReport(stmt)
           << "Iteration variable unpacking is not supported";
     }
 
-    if (targets[0].kind() != TK_VAR) {
-      throw ErrorReport(targets[0])
-          << "unexpected expression in variable initialization of for loop";
-    }
-    auto target = Var(targets[0]).name();
-
-    // match range(<expr>) style loops
-    // itrs must consist of a single Apply node
-    if (itrs[0].kind() == TK_APPLY) {
-      Apply range_iterator = Apply(itrs[0]);
-      if (range_iterator.callee().kind() == TK_VAR) {
-        Var var = Var(range_iterator.callee());
-        if (var.name().name() == "range") {
-          return emitForRange(
-              stmt.range(), target, range_iterator.inputs(), body);
-        }
-      }
-    }
-
-    // it isn't a range(<expr>) loop, treat it as a sugared value that maybe can
-    // be unrolled
+    // Emit loop information for builtinFunction values like range(), zip(), 
+    // enumerate() or SimpleValue like List, Tensor, Dict, etc.
     auto sv = emitSugaredExpr(itrs[0], 1);
 
+    // We will get IterableTree for builtinFunctions zip() and enumerate(),
+    // RangeValue for range(), and SimpleValue for types like List, Tensor, Dict.
+    auto range_val = std::dynamic_pointer_cast<RangeValue>(sv);
     auto siv = std::dynamic_pointer_cast<SimpleValue>(sv);
+    auto iterable_tree = std::dynamic_pointer_cast<IterableTree>(sv);
 
-    // for-in lists
-    if (siv && siv->getValue()->type()->kind() == TypeKind::ListType) {
-      emitForInListLoop(stmt, siv);
+    if ((siv && (siv->getValue()->type()->kind() == TypeKind::ListType
```

Comment:
These branches for different iteration kinds should be avoided. We need to fit all iteraiton into the len/getitem abstraction.